Add PR links to release post template output, and allow blog ID to be specified.

jonathansadowski opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR updates the release post tool such that it's possible to edit previous release posts. It also adds support for specifying a blog ID, if you'd like to test the tool with a different blog. Finally, it links to the PRs that introduced the changes in the templates section.

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #34248.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Make sure Docker is running and in a state to run `wp-env`
1. Checkout this branch
1. `cd` to the `tools/release-posts` directory
1. Follow the [instructions in the readme](https://github.com/woocommerce/woocommerce/blob/trunk/tools/release-posts/README.md#publishing-draft-posts) to setup for publishing release posts. If you're doing this with a test blog, use that site ID when granting access.
1. The tool will open a browser window, so be sure before running that your focused browser is logged in with access to the blog you'll use for testing.
1. Run the following: `GITHUB_ACCESS_TOKEN="..." WPCOM_OAUTH_CLIENT_ID="..." WPCOM_OAUTH_CLIENT_SECRET="..." pnpm run release-post 7.2.0 7.1.0 --siteId=...`  Be sure to provide th values for the missing variables (...). You may also choose to specify the values via your `.env` file.
1. Follow the prompts in the open browser tool to grant permissions to the site corresponding to the `siteId` that you specified.
1. Verify that the post that is created contains a link to the PR(s) that introduced the template changes.
1. Create a new post on your test site manually, add `<!-- release:templates --> This will be replaced <!-- /release:templates -->` to the source of the post you created. Note the post id.
1. Re-run the previous command with `--editPost=...`, specifying the ID of the post you just created.
1. Verify that that section of the post gets updated properly.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

jonathansadowski committed:
Update release template to be editable

jonathansadowski committed:
Correct issues in README

jonathansadowski committed:
Update to allow edits; refactor auth

jonathansadowski committed:
Update templates to add PR link

jonathansadowski committed:
Remove commented test code

psealock commented:
I keep running into this error. Does it look familiar?

```
GitError: Cloning into '/var/folders/2p/zg7bwl6d4mzb1w5m42p48yn80000gn/T/woocommerce'...
error: RPC failed; curl 56 LibreSSL SSL_read: SSL_ERROR_SYSCALL, errno 54
error: 3218 bytes of body are still expected
fetch-pack: unexpected disconnect while reading sideband packet
fatal: early EOF
fatal: fetch-pack: invalid index-pack output
```
<img width="1332" alt="image" src="https://user-images.githubusercontent.com/1922453/211222550-f328a8b7-bdf4-4608-92dc-a9e12454b2c4.png">

jonathansadowski commented:
> I keep running into this error. Does it look familiar?

Hmm. No, that one doesn't. Do you by-chance get the same error when running the command prior to these changes with options other than `--outputOnly`, or does that work for you?

psealock commented:
I'm having the same issues with just this command `pnpm run release-post contributors -- "7.2.0" "7.1.0"`

I'll need to try again with a fresh install

samueljseay commented on tools/code-analyzer/src/lib/scan-changes.ts:
```diff
@@ -22,11 +22,15 @@ export const scanForChanges = async (
+		Logger.startTask( `Making temporary clone of ${ source }...` );
+		tmpRepoPath = typeof clonedPath !== 'undefined' ? clonedPath : await cloneRepo( source );
+		Logger.endTask();
+	}
```
This is not type safe and TS complains. I think everything done here could be wrapped into the const declaration and it would be typesafe:

You don't need to worry about checking strictly for undefined.

```suggestion
	Logger.startTask( `Making temporary clone of ${ source }...` );

	const tmpRepoPath =
		typeof clonedPath !== 'undefined'
			? clonedPath
			: await cloneRepo( source );

	Logger.notice(
		`Temporary clone of ${ source } created at ${ tmpRepoPath }`
	);
```

samueljseay commented on tools/code-analyzer/src/lib/template-changes.ts:
```diff
@@ -31,8 +33,10 @@ export const scanForTemplateChanges = ( content: string, version: string ) => {
+		let lineNumber = 1;
 		let code = 'warning';
 		let message = 'This template may require a version bump!';
+		let pullRequests = [];
```
This should be a const its never re-assigned.

samueljseay commented on tools/code-analyzer/src/lib/template-changes.ts:
```diff
@@ -41,9 +45,40 @@ export const scanForTemplateChanges = ( content: string, version: string ) => {
+
+			if ( repositoryPath ) {
+				// Don't parse the headers for the patch.
+				if ( l < 4 ) {
```
TS complains since this is a string.

If it captured a string that was not a number it should probably just continue too right?

If so you could maybe do:

```
const lineNumber = parseInt(l);
if (Number.isNaN(lineNumber) || lineNumber < 4) {
...
```

samueljseay commented on tools/cli-core/src/git.ts:
```diff
@@ -190,6 +204,62 @@ export const getCommitHash = async ( baseDir: string, ref: string ) => {
+	baseDir: string,
+	filePath: string,
+	lineNumber: number
+): number => {
```
Have you got TS setup right? This has issues.

When a function is async in JS that means it returns a promise. a promise wraps the value so the type for an async function that returns a number would be Promise<number>. but also this returns a string not a number.

**also:** What's quite nice about TS is you don't need to explicitly say what the return type is and let TS infer it. In this case if you remove `: number` from the type definition TS will see that this function returns Promise<string>.

samueljseay commented on tools/cli-core/src/git.ts:
```diff
@@ -190,6 +204,62 @@ export const getCommitHash = async ( baseDir: string, ref: string ) => {
+export const getPullRequestNumberFromHash = async (
+	baseDir: string,
+	hash: string
+): number => {
```
Removing this `: number` allows TS to properly figure out the return type and not complain.

samueljseay commented on tools/release-posts/commands/release-post/release-post-release.ts:
```diff
@@ -14,11 +15,16 @@ import dotenv from 'dotenv';
+import { getWordpressComAuthToken } from '../../lib/oauth-helper';
+import { getEnvVar } from '../../lib/environment';
 import { generateContributors } from '../../lib/contributors';
+import { editPostHTML } from '../../lib/edit-post.ts';
```
This should not have the ts file extension on the import, long story short its not a valid import path.

samueljseay commented on tools/release-posts/commands/release-post/release-post-release.ts:
```diff
@@ -30,118 +36,179 @@ const program = new Command()
-			? semver.parse( options.previousVersion )
-			: semver.parse( currentVersion );
+		const siteId = options.siteId || DEVELOPER_WOOCOMMERCE_SITE_ID;
+		const tags = ( options.tags && options.tags.split( ',' ).map( ( tag ) => tag.trim() ) ) || [ 'WooCommerce Core','Releases' ];
```
as far as I can tell you've removed the tags option so this option won't ever exist? if thats intentional this should just be a const with the default array?

samueljseay commented on tools/release-posts/commands/release-post/release-post-release.ts:
```diff
@@ -30,118 +36,179 @@ const program = new Command()
+		const siteId = options.siteId || DEVELOPER_WOOCOMMERCE_SITE_ID;
+		const tags = ( options.tags && options.tags.split( ',' ).map( ( tag ) => tag.trim() ) ) || [ 'WooCommerce Core','Releases' ];
+		const isOutputOnly = !! options.outputOnly;
+		const isEditPost = !! options.editPost;
```
The way you're declaring this and then checking it before doing stuff with options.editPost, is either too smart for TS or not typesafe.

To get rid of all the warnings I would probably just remove this variable and then where you want to check for editPost just do:

`if (options.editPost) {` since an if statement in JS is falsy for: undefined, empty string etc. Perfectly fine for this case I think.

I also think probably this option should be named like editPostId or something?

samueljseay reviewed (changes requested):
I am having trouble testing the whole command locally due to internet connection so apologies for that but there are quite a lot of TS issues to resolve first so I'll retest once those are sorted.

Do you have TS properly set up in your environment because in VSCode I get lots of errors show up for these things. We're not enforcing it with any checks right now but it would be better to write the code correctly otherwise we lose all the advantages of using TypeScript.

Maybe we should punish ourselves by running TS checks for this code in PRs too though 😂

jonathansadowski committed:
Update tools/code-analyzer/src/lib/scan-changes.ts

Co-authored-by: Sam Seay <samueljseay@gmail.com>

Commit:
Address typescript issues; prettier

jonathansadowski commented:
Thanks for having a look @samueljseay. I updated this to address all the TS issues that I saw that weren't also addressed in #36412. Once that is merged, I'm happy to incorporate those changes here and making sure that that passes. If you don't mind having another look in the mean time, or I can wait for 36412 to land.

samueljseay reviewed (dismissed):
Apologies for the delay on the re-ping here. The new changes look good. 🚀

Commit:
Resolve more typescript issues

jonathansadowski commented:
Hi @samueljseay... I had to fix a few lingering issues after #36412 was merged, otherwise merging this would've caused that lint to fail. Could you please give this another look? Thanks

samueljseay reviewed (approved):
Looking good! 🚀
